src/dstack/_internal/server/services/logs.py

```diff
@@ -1,117 +1,380 @@
+        log_events_parsed = [self._runner_log_event_to_log_event(event) for event in log_events]
+        log_file_path.parent.mkdir(exist_ok=True, parents=True)
+        with open(log_file_path, "a") as f:
+            f.writelines(log.json() + "\n" for log in log_events_parsed)
+
+    def _get_log_file_path(
+        self,
+        project_name: str,
+        run_name: str,
+        job_submission_id: UUID,
+        producer: LogProducer,
+    ) -> Path:
+        return (
+            self.root
+            / "projects"
+            / project_name
+            / "logs"
+            / run_name
+            / str(job_submission_id)
+            / f"{producer.value}.log"
         )
-    logs = []
-    try:
-        with open(log_file_path) as f:
-            for line in f:
-                log_event = LogEvent.__response__.parse_raw(line)
-                if request.start_time and log_event.timestamp <= request.start_time:
-                    continue
-                if request.end_time is None or log_event.timestamp < request.end_time:
-                    logs.append(log_event)
-                else:
-                    break
-    except IOError:
-        pass
-    if request.descending:
-        logs = list(reversed(logs))
-    return JobSubmissionLogs(logs=logs)
 
+    def _runner_log_event_to_log_event(self, runner_log_event: RunnerLogEvent) -> LogEvent:
+        return LogEvent(
+            timestamp=_unix_time_ms_to_datetime(runner_log_event.timestamp),
+            log_source=LogEventSource.STDOUT,
+            message=_b64encode_raw_message(runner_log_event.message),
+        )
+
+
+def _unix_time_ms_to_datetime(unix_time_ms: int) -> datetime:
+    return datetime.fromtimestamp(unix_time_ms / 1000, tz=timezone.utc)
+
+
+def _datetime_to_unix_time_ms(dt: datetime) -> int:
+    return int(dt.timestamp() * 1000)
+
+
+def _b64encode_raw_message(message: bytes) -> str:
+    return base64.b64encode(message).decode()
+
+
+_default_log_storage: Optional[LogStorage] = None
 
-def _runner_log_event_to_log_event(runner_log_event: RunnerLogEvent) -> LogEvent:
-    return LogEvent(
-        timestamp=datetime.fromtimestamp(runner_log_event.timestamp / 1e9, tz=timezone.utc),
-        log_source=LogEventSource.STDOUT,
-        message=base64.b64encode(runner_log_event.message).decode(),
-    )
 
+def get_default_log_storage() -> LogStorage:
+    global _default_log_storage
+    if _default_log_storage is not None:
+        return _default_log_storage
+    if settings.SERVER_CLOUDWATCH_LOG_GROUP:
+        if BOTO_AVAILABLE:
+            try:
+                _default_log_storage = CloudWatchLogStorage(
+                    group=settings.SERVER_CLOUDWATCH_LOG_GROUP,
+                    region=settings.SERVER_CLOUDWATCH_LOG_REGION,
+                )
+            except LogStorageError as e:
+                logger.warning("Failed to initialize CloudWatch Logs storage: %s", e)
```

Comment:
(nit) Here and below, I'd suggest using `logger.error` to trigger Sentry. Otherwise configuration errors may be unnoticed, which may lead to logs loss